analysis/study_definition.py

```diff
@@ -1,33 +1,191 @@
+            "incidence": 0.2,
+            "date": {"earliest": "1950-01-01", "latest": "today"},
+        },
+    )
+
+def medication_count_3m_0m(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2019-12-01", "2020-02-29"],
+        returning="number_of_matches_in_period",
+        return_expectations={
+            "int": {"distribution": "normal", "mean": 3, "stddev": 2},
+            "incidence": 0.1,
+        },
+    )
+
+def medication_count_6m_3m(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2019-09-01", "2020-11-30"],
+        returning="number_of_matches_in_period",
+        return_expectations={
+            "int": {"distribution": "normal", "mean": 3, "stddev": 2},
+            "incidence": 0.1,
+        },
+    )
+
+def medication_count_12m_6m(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2019-03-01", "2020-08-31"],
+        returning="number_of_matches_in_period",
+        return_expectations={
+            "int": {"distribution": "normal", "mean": 3, "stddev": 2},
+            "incidence": 0.1,
+        },
+    )
+
+def medication_earliest(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2010-01-01", "2020-02-29"],
+        returning="date",
+        find_first_match_in_period=True,
+        include_month=True,
+        return_expectations={
+            "incidence": 0.2,
+            "date": {"earliest": "2010-01-01", "latest": "2020-02-29"},
+        },
+    )
+
+def medication_latest(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2010-01-01", "2020-02-29"],
+        returning="date",
+        find_last_match_in_period=True,
+        include_month=True,
+        return_expectations={
+            "incidence": 0.2,
+            "date": {"earliest": "2010-01-01", "latest": "2020-02-29"},
+        },
+    )
+
+def medication_counts_and_dates(var_name, med_codelist_file):
+    """
+    Generates dictionary of covariats for a medication including counts and dates
+    
+    Takes a variable prefix and medication codelist filename (minus .csv)
+    Returns a dictionary suitable for unpacking into the main study definition
+    This will include all five of the items defined in the functions above
+    """
+    
+    definitions={}
+    if (med_codelist_file[0:5] == "cross"):
+        med_codelist_file = "crossimid-codelists/" + med_codelist_file
+    else:
+        med_codelist_file = "codelists/" + med_codelist_file
+    med_codelist=codelist_from_csv(med_codelist_file + ".csv", system="snomed", column="snomed_id")
+    med_functions=[
+        ("3m_0m", medication_count_3m_0m),
+        ("6m_3m", medication_count_6m_3m),
+        ("12m_6m", medication_count_12m_6m),
+        ("earliest", medication_earliest),
+        ("latest", medication_latest)
+    ]
```

Comment:
I think the intention here is to produce 5 different columns here. There are a few things about his approach that we should reconsider
- we don't need all these for every single med
- therefore it doesn't meet principle for data minimisation
- creates extra data which can always add to run times
 - table 4 in protocol specified different dates for each one

Reply:
I've refined this to two columns for most meds and three for B cell therapies.